mailet/standard/src/main/java/org/apache/james/transport/matchers/HasHeader.java

```diff
@@ -65,9 +72,22 @@ public HeaderNameCondition(String headerName) {
         }
 
         @Override
-        public boolean isMatching(MimeMessage mimeMessage) throws MessagingException {
-            String[] headerArray = mimeMessage.getHeader(headerName);
-            return headerArray != null && headerArray.length > 0;
+        public Collection<MailAddress> isMatching(Mail mail) throws MessagingException {
+            String[] headerArray = mail.getMessage().getHeader(headerName);
+            if (headerArray != null && headerArray.length > 0) {
+                return mail.getRecipients();
+            }
+
+            return matchSpecific(mail);
+        }
+
+        protected Collection<MailAddress> matchSpecific(Mail mail) {
+            return mail.getPerRecipientSpecificHeaders().getHeadersByRecipient() //
+                    .entries() //
+                    .stream() //
+                    .filter(entry -> entry.getValue().getName().equals(headerName)) //
+                    .map(entry -> entry.getKey()) //
```

Comment:
I see those empty comments // a bit everywhere in your code. Any reason?

Reply:
to force my code formatter in Eclipse to create a new line.
let me remove that since your project is not doing that